docs/RFCS/20211106_multitenant_cluster_settings.md

```diff
@@ -0,0 +1,274 @@
+
+ - We add a new `system.tenant_settings` table with following schema:
+   ```
+   CREATE TABLE system.tenant_settings (
+     tenant_id INT8 NOT NULL,
+     name STRING NOT NULL,
+     value STRING NOT NULL,
+     value_type STRING,
+     last_updated TIMESTAMP NOT NULL DEFAULT now() ON UPDATE now(),
+     reason STRING,
+     PRIMARY KEY (tenant_id, name)
+   )
+   ```
+   This table is only used on the system tenant. All-non-system-tenant values
+   are stored in `tenant_id=0`. This table contains no settings for the system
+   tenant (`tenant_id=1`), and the `tenant_id=0` values do not apply to the
+   system tenant.
+
+ - We modify the tenant connector APIs to allow "listening" for updates to
+   cluster settings. Inside the tenant connector this can be implemented using a
+   streaming RPC (similar to `GossipSubscription`).
+
+ - On the system tenant we set up rangefeed on `system.tenant_settings` and keep
+   all the changed settings (for all tenants) in memory. We expect that in
+   practice overrides for specific tenants are rare (with most being "all
+   tenant" overrides). The rangefeed is used to implement the API used by the
+   tenant connector to keep tenants up to date. We continue to set up the
+   rangefeed on the `system.settings` table to maintain the system tenant
+   settings.
+
+ - On non-system tenants we continue to set up the rangefeed on the tenant's
+   `system.settings` table, and we also use the new connector API to listen to
+   updates from the host. Values from the host which are present always override
+   any local values.
+
+
+### Upgrade
+
+The proposed solution has very few concerns around upgrade. There will be a
+migration to create the new system table, and the new connector API
+implementation is only active on the new version (in a mixed-version cluster, it
+can error out or use a stub no-op implementation). The new statements (around
+setting per-tenant values) should also error out until the new version is
+active.
+
+The settings on the system tenant will continue to work. On non-system tenants,
+any locally changed settings that are now `system` or `tenant-rw` will revert to
+their defaults. It will be necessary to set these settings from the system
+tenant (using the new statements) if any clusters rely on non-default values.
+
+### Notes
+
+All functions used to register cluster settings take an extra argument with the
+class of the setting. We want to make an explicit (and reviewable) decision for
+each existing cluster setting, and we want the authors of future settings to be
+forced to think about the class.
+
+When deciding which class is appropriate for a given setting, we will use the
+following guidelines:
+
+ - if the setting controls a user-visible aspect of SQL, it should be a
+   `tenant-rw` setting.
+
+ - control settings relevant to tenant-specific internal implementation (like
+   tenant throttling) that we want to be able to control per-tenant should be
```

Comment:
Will `system` settings need to be controlled per tenant? I thought they only applied to the host cluster